protocols/jvss/jvss.go

```diff
@@ -15,645 +21,249 @@ import (
 	for i, tn := range nodes {
-		if tn.Id.Equals(n.TreeNode().Id) {
-			idx = i
+		pk[i] = tn.Entity.Public
+	}
+	// NOTE: T <= R <= N (for simplicity we use T = R = N; might change later)
+	info := poly.Threshold{T: len(nodes), R: len(nodes), N: len(nodes)}
+
+	jv := &JVSS{
+		Node:        node,
+		keyPair:     kp,
+		nodeList:    nodes,
+		pubKeys:     pk,
+		info:        info,
+		schnorr:     new(poly.Schnorr),
+		secrets:     make(map[string]*Secret),
+		ltssInit:    false,
+		secretsDone: make(chan bool, 1),
+		sigChan:     make(chan *poly.SchnorrSig),
+	}
+
+	// Setup message handlers
+	handlers := []interface{}{
+		jv.handleSecInit,
+		jv.handleSecConf,
+		jv.handleSigReq,
+		jv.handleSigResp,
+	}
+	for _, h := range handlers {
+		if err := jv.RegisterHandler(h); err != nil {
+			return nil, fmt.Errorf("Error, could not register handler: " + err.Error())
 		}
-		pubs[i] = tn.Entity.Public
-	}
-	if idx == -1 {
-		panic(fmt.Sprintf("Could not find JVSS node %+v in the list of nodes %+v", n, nodes))
 	}
-	kp := config.KeyPair{Public: n.Entity().Public, Secret: n.Private(), Suite: n.Suite()}
-	nbPeers := len(tree.EntityList.List)
-	info := poly.Threshold{T: nbPeers, R: nbPeers, N: nbPeers}
-	jv := &JVSSProtocol{
-		Node:         n,
-		List:         tree.EntityList,
-		index:        idx,
-		info:         info,
-		publicList:   pubs,
-		key:          kp,
-		nodeList:     nodes,
-		requests:     make(map[int]*RequestBuffer),
-		ltChan:       make(chan LongtermChan),
-		rdChan:       make(chan RandomChan),
-		reqChan:      make(chan RequestChan),
-		respChan:     make(chan ResponseChan),
-		longtermLock: new(sync.Mutex),
-		longterm:     NewLongtermRequest(n.Suite(), info, kp),
-	}
-	if err := jv.Node.RegisterChannel(jv.ltChan); err != nil {
-		return nil, err
-	}
-	if err := jv.Node.RegisterChannel(jv.rdChan); err != nil {
-		return nil, err
-	}
-	if err := jv.Node.RegisterChannel(jv.reqChan); err != nil {
-		return nil, err
-	}
-	if err := jv.Node.RegisterChannel(jv.respChan); err != nil {
-		return nil, err
-	}
-	// FIXME leaky go rountines in tests (only?)
-	go jv.waitForRandom()
-	go jv.waitForResponses()
-	go jv.waitForRequests()
-	go jv.waitForLongterm()
-	return jv, nil
-}
 
-func NewJVSSProtocolInstance(node *sda.Node) (sda.ProtocolInstance, error) {
-	return NewJVSSProtocol(node)
-}
-
-// Start will send the message to first compute the long term secret
-// It's a blocking call because we are supposed to launch that into a go
-// routine anyway from sda.
-func (jv *JVSSProtocol) Start() error {
-	jv.waitLongtermSecret()
-	return nil
+	return jv, nil
 }
 
-func (jv *JVSSProtocol) Shutdown() error {
-	close(jv.reqChan)
-	close(jv.respChan)
-	close(jv.ltChan)
-	close(jv.rdChan)
+// Start initiates the JVSS protocol by setting up a long-term shared secret
+// which can be used later on by the JVSS group to sign and verify messages.
+func (jv *JVSS) Start() error {
+	jv.initSecret(LTSS)
+	<-jv.secretsDone
```

Comment:
The normal `Start` method doesn't wait for the protocol to finish. @nikkolasg , do you think it is OK to 'delete' the `jv.secretsDone` here?

You can either make `secretsDone` public (but then you have to make sure it's empty before you write something into it, else you'll block), make a `WaitForDone` method or do a `RegisterDoneCallback`.